File: tests/test_normal.cpp
// Unit tests for Normal distribution
#include <cassert>
#include <cmath>
#include <iostream>

#include <fastdist/math/normal.h>

void test_normal() {
    std::cout << "Running normal distribution tests...\n";

    constexpr double tol = 1e-12;
    constexpr double mu = 1.5;
    constexpr double sigma = 2.0;

    // -------------------------
    // PDF / logPDF / CDF tests
    // -------------------------
    {
        const double pdf0 = fastdist::math::normal_pdf_scalar(mu, mu, sigma);
        const double expected_pdf = 1.0 / (sigma * std::sqrt(2.0 * M_PI));
        assert(std::abs(pdf0 - expected_pdf) < tol);

        const double logpdf0 = fastdist::math::normal_logpdf_scalar(mu, mu, sigma);
        assert(std::abs(std::exp(logpdf0) - expected_pdf) < tol);

        const double cdf0 = fastdist::math::normal_cdf_scalar(mu, mu, sigma);
        assert(std::abs(cdf0 - 0.5) < tol);
    }

    // -------------------------
    // Mean / Variance / Stddev
    // -------------------------
    {
        assert(std::abs(fastdist::math::normal_mean(mu) - mu) < tol);
        assert(std::abs(fastdist::math::normal_variance(sigma) - sigma * sigma) < tol);
        assert(std::abs(fastdist::math::normal_stddev(sigma) - sigma) < tol);
    }

    // -------------------------
    // Z-score
    // -------------------------
    {
        const double z = fastdist::math::z_score(mu + sigma, mu, sigma);
        assert(std::abs(z - 1.0) < tol);
    }

    // -------------------------
    // MGF tests
    // M(t) = exp(mu t + 0.5 sigma^2 t^2)
    // -------------------------
    {
        const double mgf0 = fastdist::math::normal_mgf_scalar(0.0, mu, sigma);
        assert(std::abs(mgf0 - 1.0) < tol);

        constexpr double t = 0.25;
        const double expected = std::exp(mu * t + 0.5 * sigma * sigma * t * t);

        const double mgf = fastdist::math::normal_mgf_scalar(t, mu, sigma);
        assert(std::abs(mgf - expected) < tol);

        // M'(0) = mean
        constexpr double dt = 1e-6;
        const double dM =
                (fastdist::math::normal_mgf_scalar(dt, mu, sigma) - fastdist::math::normal_mgf_scalar(-dt, mu, sigma)) /
                (2.0 * dt);

        assert(std::abs(dM - mu) < 1e-6);
    }

    // -------------------------
    // CGF tests
    // K(t) = mu t + 0.5 sigma^2 t^2
    // -------------------------
    {
        const double cgf0 = fastdist::math::normal_cgf_scalar(0.0, mu, sigma);
        assert(std::abs(cgf0) < tol);

        const double t = 0.25;
        const double expected = mu * t + 0.5 * sigma * sigma * t * t;

        const double cgf = fastdist::math::normal_cgf_scalar(t, mu, sigma);
        assert(std::abs(cgf - expected) < tol);

        // K'(0) = mean
        constexpr double dt = 1e-6;
        const double dK =
                (fastdist::math::normal_cgf_scalar(dt, mu, sigma) - fastdist::math::normal_cgf_scalar(-dt, mu, sigma)) /
                (2.0 * dt);

        assert(std::abs(dK - mu) < 1e-6);

        // K''(0) = variance
        const double d2K = (fastdist::math::normal_cgf_scalar(dt, mu, sigma) -
                            2.0 * fastdist::math::normal_cgf_scalar(0.0, mu, sigma) +
                            fastdist::math::normal_cgf_scalar(-dt, mu, sigma)) /
                           (dt * dt);

        assert(std::abs(d2K - sigma * sigma) < 1e-6);
    }

    // -------------------------
    // Normal RNG tests
    // -------------------------
    {
        constexpr int N = 200000;
        double sum = 0.0;
        double sumsq = 0.0;

        for (int i = 0; i < N; ++i) {
            const double x = fastdist::math::normal_sample(mu, sigma);

            sum += x;
            sumsq += x * x;
        }

        const double mean = sum / N;
        const double var = sumsq / N - mean * mean;

        assert(std::abs(mean - mu) < 5e-2);
        assert(std::abs(var - sigma * sigma) < 5e-2);
    }

    // -------------------------
    // Log-normal RNG tests
    // -------------------------
    {
        constexpr int N = 250000;
        double sum = 0.0;

        for (int i = 0; i < N; ++i) {
            const double x = fastdist::math::normal_log_sample(mu, sigma);
            assert(x > 0.0);
            sum += std::log(x);
        }

        const double log_mean = sum / N;
        assert(std::abs(log_mean - mu) < 5e-2);
    }

    std::cout << "Normal distribution tests passed.\n";
}

File: tests/python/test_normal.py
# tests/test_normal.py
import math

import numpy as np
import pytest

from fastdist.distributions.normal import Normal


# ---------------------------
# Fixtures
# ---------------------------
@pytest.fixture
def standard_normal():
    return Normal(mu=0, sigma=1)


@pytest.fixture
def shifted_normal():
    return Normal(mu=5, sigma=2)


# ---------------------------
# Constructor & Properties
# ---------------------------
def test_constructor_valid(standard_normal):
    assert standard_normal.mu == 0.0
    assert standard_normal.sigma == 1.0


def test_constructor_invalid_mu():
    with pytest.raises(TypeError):
        Normal(mu="invalid", sigma=1)
    with pytest.raises(ValueError):
        Normal(mu=float('inf'), sigma=1)


def test_constructor_invalid_sigma():
    with pytest.raises(TypeError):
        Normal(mu=0, sigma="invalid")
    with pytest.raises(ValueError):
        Normal(mu=0, sigma=0)
    with pytest.raises(ValueError):
        Normal(mu=0, sigma=-1)
    with pytest.raises(ValueError):
        Normal(mu=0, sigma=float('nan'))


def test_setters(standard_normal):
    standard_normal.mu = 2
    assert standard_normal.mu == 2.0
    standard_normal.sigma = 3
    assert standard_normal.sigma == 3.0
    with pytest.raises(ValueError):
        standard_normal.sigma = 0


# ---------------------------
# PDF / logPDF / CDF
# ---------------------------
def test_pdf_scalar(standard_normal):
    x = 0
    pdf_val = standard_normal.pdf(x)
    expected = 1 / math.sqrt(2 * math.pi)
    assert math.isclose(pdf_val, expected, rel_tol=1e-9)


def test_logpdf_scalar(standard_normal):
    x = 0
    logpdf_val = standard_normal.logpdf(x)
    expected = math.log(1 / math.sqrt(2 * math.pi))
    assert math.isclose(logpdf_val, expected, rel_tol=1e-9)


def test_cdf_scalar(standard_normal):
    x = 0
    cdf_val = standard_normal.cdf(x)
    assert math.isclose(cdf_val, 0.5, rel_tol=1e-9)


def test_pdf_vector(standard_normal):
    x = [-1, 0, 1]
    result = standard_normal.pdf(x)
    assert isinstance(result, np.ndarray)
    assert result.shape == (3,)
    assert math.isclose(result[1], 1 / math.sqrt(2 * math.pi), rel_tol=1e-9)


# ---------------------------
# MGF / CGF
# ---------------------------
def test_mgf_cgf_scalar(shifted_normal):
    t = 1.0
    mgf_val = shifted_normal.mgf(t)
    expected_mgf = math.exp(shifted_normal.mu * t + 0.5 * shifted_normal.sigma ** 2 * t ** 2)
    assert math.isclose(mgf_val, expected_mgf, rel_tol=1e-9)

    cgf_val = shifted_normal.cgf(t)
    expected_cgf = math.log(expected_mgf)
    assert math.isclose(cgf_val, expected_cgf, rel_tol=1e-9)


# ---------------------------
# Sampling
# ---------------------------
def test_sample(standard_normal):
    val = standard_normal.sample()
    assert isinstance(val, float)


def test_log_sample(standard_normal):
    val = standard_normal.log_sample()
    assert isinstance(val, float)


# ---------------------------
# Z-score
# ---------------------------
def test_z_score_scalar(standard_normal):
    z = standard_normal.z_score(1)
    assert math.isclose(z, 1.0)


# ---------------------------
# CUDA availability
# ---------------------------
def test_cuda_availability():
    available = Normal.is_cuda_available()
    assert isinstance(available, bool)


# ---------------------------
# Scalar class methods
# ---------------------------
def test_pdf_scalar_class_method():
    val = Normal._pdf_scalar(0, mu=0, sigma=1)
    expected = 1 / math.sqrt(2 * math.pi)
    assert math.isclose(val, expected, rel_tol=1e-9)


def test_logpdf_scalar_class_method():
    val = Normal._logpdf_scalar(0, mu=0, sigma=1)
    expected = math.log(1 / math.sqrt(2 * math.pi))
    assert math.isclose(val, expected, rel_tol=1e-9)


def test_cdf_scalar_class_method():
    val = Normal._cdf_scalar(0, mu=0, sigma=1)
    assert math.isclose(val, 0.5, rel_tol=1e-9)


def test_logpdf_accepts_an_array(standard_normal):
    """
    The vectorised path once dropped step_size when forwarding to the backend,
    which raised TypeError for every array input. Only the scalar path was
    covered, so nothing caught it.
    """
    x = np.array([-1.0, 0.0, 1.0])
    got = standard_normal.logpdf(x)
    expected = [math.log(1.0 / math.sqrt(2.0 * math.pi)) - v * v / 2.0 for v in x]
    assert np.allclose(got, expected)


# ---------------------------------------------------------------------------
# Lower-tail precision
#
# 0.5 * (1 + erf(z)) cancels to nothing far below the mean, so the CDF lost
# all relative precision below about -8 sigma and returned exactly 0 at -10.
# Tail probabilities are what p-values are made of, so they are checked here
# against erfc directly, on both the scalar and the batch path.
# ---------------------------------------------------------------------------

def _phi(x):
    return 0.5 * math.erfc(-x / math.sqrt(2.0))


@pytest.mark.parametrize("x", [-5.0, -8.0, -10.0, -20.0, -37.0])
def test_cdf_keeps_relative_precision_in_the_lower_tail(x):
    import fastdist._fastdist as core
    expected = _phi(x)
    assert core.normal_cdf_scalar(x, 0.0, 1.0) == pytest.approx(expected, rel=1e-12)
    batch = core.normal_cdf_cpu(np.array([x]), 0.0, 1.0, 0.0)
    assert batch[0] == pytest.approx(expected, rel=1e-12)
